kythe/docs/schema/indexing-protobuf.txt

```diff
@@ -31,12 +31,235 @@ implementation, as this is often difficult (or uninteresting) to read. This
+        "type": "anchor_anchor",
+        "source_begin": 13,
+        "source_end": 18,
+        "target_begin": 6,
+        "target_end": 11,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }, {
+        "type": "anchor_anchor",
+        "source_begin": 22,
+        "source_end": 25,
+        "target_begin": 18,
+        "target_end": 21,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }]
+}
+-------------------------------------------------------------------------------
+
+This mapping has 3 `meta` entries. The first entry for `Color`, the second for
+`RED`. Note:
+
+* Each entry doesn't contain names of elements. Each entry contains only
+  position of elements in the source (`Color.java`) and target (`color.js`)
+  files.
+* Each position is defined as byte offset inside file and not as line/column.
+  This is required because in Kythe anchors defined using byte offsets and
+  not line/column. In this example JavaScript indexer will process this
+  mapping and will need to output *anchor* for `Color.java` and indexer
+  doesn't have access to the `Color.java` file (it has access only to JS
+  files). Because of that JS indexer can't translate line/column to byte
+  offset.
+* Entry doesn't contain vnames of elements in `Color.java` or `color.js` and
+  instead contains positions. Vnames of nodes are internal details of each
+  indexer and subject to change. Generator usually a standalone tool that
+  doesn't know rules for producing vnames for specific language so it's
+  impossible for Generator to output vnames of nodes.
+
+To pass this mapping to the JavaScript Indexer Generator will append it
+as a comment at the last line of `color.js`:
+
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+
+// Kythe Indexing Metadata:
+// {"type":"kythe0","meta":[{"type":"anchor_anchor","source_begin":13,...
+-------------------------------------------------------------------------------
+
+Inlining metadata inside `color.js` has benefit of avoiding passing extra
+files to Indexer. All Indexer needs is to know that some JavaScript files can
+contain metadata on the last line and parse it.
+
+One downside is that it adds noise to `color.js` but usually generated
+files are invisible to developers so it's not a big concern.
+
+==== Changes to JavaScript Indexer
+
+On JavaScript Indexer side we need to parse metadata and output *imputes*
+edges. To parse metadata indexer can check last two lines of all `.js` files
+and see if they contain `// Kythe Indexind Metadata:` and if so - parse
+the last line as JSON.
+
+For each `meta` entry indexer should do the following:
+
+1. Output an *anchor* using `source_begin` and `source_end`. `source_vname`
```

Comment:
This varies by edge and "type", e.g. kythe/cxx/indexer/cxx/testdata/metadata/inline.meta

Where a reverse-generates edge should be painted from the node found at that location to the semantic node designated by the vname.

Reply:
Also looking at the above, it appears as though JS and C++ might use slightly different JSON metadata? That would be unfortunate...

Reply:
1. This approach is talking about `imputes` edges, not `generates`. In my mental picture those are two different approaches based on availability of VNames. So in this doc I'm trying to keep them separate. If that's not the right mental picture - how should I change it?

2. What is that C++ JSON metadata used for? I was only aware of `GeneratedCodeInfo` binary proto used by protobuf <=> C++. And yes, this JSON and the one I described seem to have similar fields but different structure :/ Maybe I can unify them but need to understand better C++ use case.

Reply:
That's the right mental picture.  But from a metadata perspective, they're similar. Same application rules, but different "payload".

I think that JSON metadata format is used by Mojom (a Chromium thing similar to proto). I filed #3711 to document/specify the formats.  Looking a bit more closely, it appears the structure varies a lot by the "type" field.  The "anchor_anchor" type is the same(?) in both JavaScript and C++, but C++ also supports "anchor_defines", which has a different structure.

Reply:
I think for here, it's enough to note that different "type" fields have a different expected structure and here we're using "anchor_anchor" which is for imputes edges.